src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs

```diff
@@ -489,13 +496,41 @@ internal bool EntryIsFresh(ResponseHeaders responseHeaders, TimeSpan age, bool v
             if (!responseServed && RequestCacheControl.OnlyIfCached)
             {
                 _httpContext.Response.StatusCode = StatusCodes.Status504GatewayTimeout;
-
-                responseServed = true;
+                return true;
             }
 
             return responseServed;
         }
 
+        internal bool ConditionalRequestSatisfied(ResponseHeaders cachedResponseHeaders)
+        {
+            if (RequestHeaders.IfNoneMatch != null)
```

Comment:
This property parses every time you access it, with no caching. Put it in a local variable